doc/SAI-Proposal-HW-Resource.md

```diff
@@ -0,0 +1,167 @@
+SAI Infra for distributed HW resource management
+-------------------------------------------------------------------------------
+ Title       | SAI distributed HW resource management
+-------------|-----------------------------------------------------------------
+ Authors     | Jai Kumar, Broadcom Inc.
+ Status      | In review
+ Type        | Standards track
+ Created     | 07/08/2022: Initial Draft
+ SAI-Version | 1.9
+-------------------------------------------------------------------------------
+
+Modern chips have distributed architecture where HW resources are split between ingress and egress stages of the pipleine for scaling and performance reasons.
+
+PR #1482 introduces a method to
+1. Specify the stage of a resource
+2. Query an object's attribute stage for a given HW using capability API
+3. Error handling for incorrect resource stage specification
+
+## Common Data Type
+-----------------
+New data type sai_object_stage_t is introduced. This data type specifies if a resource is of type
+
+- SAI_OBJECT_STAGE_BOTH:
+  Resource is a common shared resource between ingress and egress HW pipeline
+- SAI_OBJECT_STAGE_INGRESS:
+  Resource belongs to ingress stage of HW pipeline
+- SAI_OBJECT_STAGE_EGRESS:
+  Resource belongs to egress stage of HW pipeline
+
+## Capability Query
+----------------
+New  query API is introduced for an attribute's stage. NOS can query the stage of an object's attribute for a given HW support. One of the following stage is returned as part the API
+- SAI_ATTR_STAGE_NA: Stage is not applicable
+- SAI_ATTR_STAGE_BOTH: Attribute is common to ingress and egress stage
+- SAI_ATTR_STAGE_INGRESS: Attribute is applicable only to ingress stage
+- SAI_ATTR_STAGE_EGRESS: Attribute is applicabe only to egress stage
+
+Length of the returned array "stage" is attr_count. Caller must provide the buffer for array "stage".

```

Comment:
This is not applicable. The output will be a pointer to a single sai_object_stage_t struct.

Reply:
Please see the update workflow document. It will be similar to attr_list argument with attr_count except in return direction where attr_count indicates the size of the stage array.

Reply:
Jai, Can we discuss this in the upcoming call? The confusion is can the attribute_id be passed multiple times with difference values in a single call to sai_query_object_stage(). The workflow example does this but it looks a little confusing.  The intent of passing an array of sai_attribute_t to the API was since the called would not know which attribute-id decides the stage for a given vendor...